README.rst
Channels API
------------

.. image:: https://travis-ci.org/linuxlewis/channels-api.svg?branch=master
    :target: https://travis-ci.org/linuxlewis/channels-api

Channels API exposes a RESTful Streaming API over WebSockets using
channels. It provides a ``ResourceBinding`` which is comparable to Django
Rest Framework's ``ModelViewSet``. It is based on DRF serializer
classes.

It requires Python 2.7 or 3.x, Channels <=1.1.8.1, Django <=1.11, and Django Rest Framework 3.x

You can learn more about channels-api from my talk at the `SF Django Meetup <https://vimeo.com/194110172#t=3033>`__ or `PyBay 2016 <https://www.youtube.com/watch?v=HzC_pUhoW0I>`__

Table of Contents
-----------------

-  `Getting Started <#getting-started>`__
-  `ResourceBinding <#resourcebinding>`__
-  `Subscriptions <#subscriptions>`__
-  `Custom Actions <#custom-actions>`__
-  `Permissions <#permissions>`__


How does it work?
-----------------

The API builds on top of channels' ``WebsocketBinding`` class. It works by having
the client send a ``stream`` and ``payload`` parameters. This allows
us to route messages to different streams (or resources) for a particular
action. So ``POST /user`` would have a message that looks like the following

.. code:: javascript

    var msg = {
      stream: "users",
      payload: {
        action: "create",
        data: {
          email: "test@example.com",
          password: "password"
        }
      }
    }

    ws.send(JSON.stringify(msg))

Why?
----

You're already using Django Rest Framework and want to expose similar
logic over WebSockets.

WebSockets can publish updates to clients without a request. This is
helpful when a resource can be edited by multiple users across many platforms.

Getting Started
---------------

This tutorial assumes you're familiar with channels and have completed
the `Getting
Started <https://channels.readthedocs.io/en/latest/getting-started.html>`__

-  Add ``channels_api`` to requirements.txt

.. code:: bash

  pip install channels_api

-  Add ``channels_api`` to ``INSTALLED_APPS``

.. code:: python


    INSTALLED_APPS = (
        'rest_framework',
        'channels',
        'channels_api'
    )

-  Add your first resource binding

.. code:: python


    # polls/bindings.py

    from channels_api.bindings import ResourceBinding

    from .models import Question
    from .serializers import QuestionSerializer

    class QuestionBinding(ResourceBinding):

        model = Question
        stream = "questions"
        serializer_class = QuestionSerializer
        queryset = Question.objects.all()

-  Add a ``WebsocketDemultiplexer`` to your ``channel_routing``

.. code:: python

    # proj/routing.py


    from channels.generic.websockets import WebsocketDemultiplexer
    from channels.routing import route_class

    from polls.bindings import QuestionBinding

    class APIDemultiplexer(WebsocketDemultiplexer):

        consumers = {
          'questions': QuestionBinding.consumer
        }

    channel_routing = [
        route_class(APIDemultiplexer)
    ]

That's it. You can now make REST WebSocket requests to the server.

.. code:: javascript

    var ws = new WebSocket("ws://" + window.location.host + "/")

    ws.onmessage = function(e){
        console.log(e.data)
    }

    var msg = {
      stream: "questions",
      payload: {
        action: "create",
        data: {
          question_text: "What is your favorite python package?"
        },
        request_id: "some-guid"
      }
    }
    ws.send(JSON.stringify(msg))
    // response
    {
      stream: "questions",
      payload: {
        action: "create",
        data: {
          id: "1",
          question_text: "What is your favorite python package"
        }
        errors: [],
        response_status: 200
        request_id: "some-guid"
      }
    }

-  Add the channels debugger page (Optional)

This page is helpful to debug API requests from the browser and see the
response. It is only designed to be used when ``DEBUG=TRUE``.

.. code:: python

    # proj/urls.py

    from django.conf.urls import include

        urlpatterns = [
            url(r'^channels-api/', include('channels_api.urls'))
        ]

ResourceBinding
---------------

By default the ``ResourceBinding`` implements the following REST methods:

- ``create``
- ``retrieve``
- ``update``
- ``list``
- ``delete``
- ``subscribe``

See the test suite for usage examples for each method.


List Pagination
---------------

Pagination is handled by `django.core.paginator.Paginator`

You can configure the ``DEFAULT_PAGE_SIZE`` by overriding the settings.


.. code:: python

  # settings.py

  CHANNELS_API = {
    'DEFAULT_PAGE_SIZE': 25
  }


Subscriptions
-------------

Subscriptions are a way to programmatically receive updates
from the server whenever a resource is created, updated, or deleted

By default channels-api has implemented the following subscriptions

- create a Resource
- update any Resource
- update this Resource
- delete any Resource
- delete this Resource

To subscribe to a particular event just use the subscribe action
with the parameters to filter

.. code:: javascript

  // get an event when any question is updated

  var msg = {
    stream: "questions",
    payload: {
      action: "subscribe",
      data: {
        action: "update"
      }
    }
  }

  // get an event when question(1) is updated
  var msg = {
    stream: "questions",
    payload: {
      action: "subscribe",
      pk: "1",
      data: {
        action: "update"
      }
    }
  }


Custom Actions
--------------

To add your own custom actions, use the ``detail_action`` or ``list_action``
decorators.


.. code:: python

    from channels_api.bindings import ResourceBinding
    from channels_api.decorators import detail_action, list_action

    from .models import Question
    from .serializers import QuestionSerializer

    class QuestionBinding(ResourceBinding):

        model = Question
        stream = "questions"
        serializer_class = QuestionSerializer
        queryset = Question.objects.all()

        @detail_action()
        def publish(self, pk, data=None, **kwargs):
            instance = self.get_object(pk)
            result = instance.publish()
            return result, 200

        @list_action()
        def report(self, data=None, **kwargs):
            report = self.get_queryset().build_report()
            return report, 200

Then pass the method name as "action" in your message

.. code:: javascript

  // run the publish() custom action on Question 1
  var msg = {
    stream: "questions",
    payload: {
      action: "publish",
      data: {
        pk: "1"
      }
    }
  }

  // run the report() custom action on all Questions
  var msg = {
    stream: "questions",
    payload: {
      action: "report"
    }
  }

Permissions
-----------

Channels API offers a simple permission class system inspired by rest_framework.
There are two provided permission classes: ``AllowAny`` and ``IsAuthenticated``.

To configure permissions globally use the setting ``DEFAULT_PERMISSION_CLASSES`` like so

.. code:: python

    # settings.py

    CHANNELS_API = {
        'DEFAULT_PERMISSION_CLASSES': ('channels_api.permissions.AllowAny',)

    }

You can also configure the permission classes on a ``ResourceBinding`` itself like so

.. code:: python

    from channels_api.permissions import IsAuthenticated

    class MyBinding(ResourceBinding):
        permission_classes = (IsAuthenticated,)


Lastly, to implement your own permission class, override the ``has_permission`` of ``BasePermission``.

.. code:: python

    from channels_api.permissions import BasePermission

    class MyPermission(BasePermission):

        def has_permission(self, user, action, pk):

            if action == "CREATE":
                return True
            return False


CHANGELOG
0.4.1 - Released March 4th 2018
---
- Pinned dependences to channels 1.x
- Added partial update



tests/models.py
from django.db import models

class TestModel(models.Model):
    """Simple model to test with."""

    name = models.CharField(max_length=255)

tests/routing.py
from channels import route_class
from channels_api.routing import Routing
from .test_bindings import TestModelResourceBinding, TestPermissionResourceBinding


class TestDemultiplexer(Routing):

    consumers = {
        'testmodel': TestModelResourceBinding.consumer,
        'user:testmodel': TestPermissionResourceBinding.consumer,
    }

channel_routing = [
    route_class(TestDemultiplexer)
]

tests/test_bindings.py
import json

from django.utils.encoding import force_text
from django.contrib.auth.models import User
from rest_framework import serializers

from channels.message import pending_message_store
from channels.tests import ChannelTestCase, Client

from channels_api import bindings, permissions
from channels_api.settings import api_settings

from .models import TestModel


class TestModelSerializer(serializers.ModelSerializer):
    class Meta:
        model = TestModel
        fields = ('id', 'name')


class TestModelResourceBinding(bindings.ResourceBinding):

    model = TestModel
    queryset = TestModel.objects.all()
    serializer_class = TestModelSerializer
    stream = 'testmodel'


class TestPermissionResourceBinding(bindings.ResourceBinding):

    model = TestModel
    queryset = TestModel.objects.all()
    serializer_class = TestModelSerializer
    permission_classes = (permissions.IsAuthenticated, )
    stream = 'testmodel'


class ResourceBindingTestCase(ChannelTestCase):

    def setUp(self):
        super(ResourceBindingTestCase, self).setUp()
        self.client = Client()

    def _send_and_consume(self, channel, data):
        """Helper that sends and consumes message and returns the next message."""
        self.client.send_and_consume(force_text(channel), data)
        return self._get_next_message()

    def _get_next_message(self):
        msg = self.client.get_next_message(self.client.reply_channel)
        try:
            return json.loads(msg['text'])
        except Exception:
            return None

    def _build_message(self, stream, payload):
        return {"text": json.dumps({"stream": stream, "payload": payload}), "path": "/"}

    def _build_authentication(self, username, password):
        return {"username": username, "password": password, "text": "", "path": "/"}

    def test_create(self):
        """Integration that asserts routing a message to the create channel.

        Asserts response is correct and an object is created.
        """
        json_content = self._send_and_consume("websocket.receive", self._build_message("testmodel", {
            'action': 'create',
            'pk': None,
            'request_id': 'client-request-id',
            'data': {'name': 'some-thing'}
        }))
        # it should create an object
        self.assertEqual(TestModel.objects.count(), 1)

        expected = {
            'action': 'create',
            'data': TestModelSerializer(TestModel.objects.first()).data,
            'errors': [],
            'request_id': 'client-request-id',
            'response_status': 201
        }
        # it should respond with the serializer.data
        self.assertEqual(json_content['payload'], expected)

    def test_create_failure(self):
        """Integration that asserts error handling of a message to the create channel."""

        json_content = self._send_and_consume('websocket.receive', self._build_message("testmodel", {
            'action': 'create',
            'pk': None,
            'request_id': 'client-request-id',
            'data': {},
        }))
        # it should not create an object
        self.assertEqual(TestModel.objects.count(), 0)

        expected = {
            'action': 'create',
            'data': None,
            'request_id': 'client-request-id',
            'errors': [{'name': ['This field is required.']}],
            'response_status': 400
        }
        # it should respond with an error
        self.assertEqual(json_content['payload'], expected)

    def test_delete(self):
        instance = TestModel.objects.create(name='test-name')

        json_content = self._send_and_consume('websocket.receive', self._build_message('testmodel', {
            'action': 'delete',
            'pk': instance.id,
            'request_id': 'client-request-id',
        }))

        expected = {
            'action': 'delete',
            'errors': [],
            'data': {},
            'request_id': 'client-request-id',
            'response_status': 200
        }
        self.assertEqual(json_content['payload'], expected)
        self.assertEqual(TestModel.objects.count(), 0)

    def test_delete_failure(self):
        json_content = self._send_and_consume('websocket.receive', self._build_message('testmodel', {
            'action': 'delete',
            'pk': -1,
            'request_id': 'client-request-id'
        }))

        expected = {
            'action': 'delete',
            'errors': ['Not found.'],
            'data': None,
            'request_id': 'client-request-id',
            'response_status': 404
        }

        self.assertEqual(json_content['payload'], expected)

    def test_list(self):
        for n in range(api_settings.DEFAULT_PAGE_SIZE + 1):
            TestModel.objects.create(name='Name-{}'.format(str(n)))

        json_content = self._send_and_consume('websocket.receive', self._build_message('testmodel', {
            'action': 'list',
            'request_id': 'client-request-id',
            'data': None,
        }))

        self.assertEqual(len(json_content['payload']['data']), api_settings.DEFAULT_PAGE_SIZE)

        json_content = self._send_and_consume('websocket.receive', self._build_message('testmodel', {
            'action': 'list',
            'request_id': 'client-request-id',
            'data': {
                'page': 2
            }
        }))

        self.assertEqual(len(json_content['payload']['data']), 1)
        self.assertEqual('client-request-id', json_content['payload']['request_id'])

    def test_retrieve(self):

        instance = TestModel.objects.create(name="Test")

        json_content = self._send_and_consume('websocket.receive', self._build_message('testmodel', {
            'action': 'retrieve',
            'pk': instance.id,
            'request_id': 'client-request-id'
        }))
        expected = {
            'action': 'retrieve',
            'data': TestModelSerializer(instance).data,
            'errors': [],
            'response_status': 200,
            'request_id': 'client-request-id'
        }
        self.assertTrue(json_content['payload'] == expected)

    def test_retrieve_404(self):
        json_content = self._send_and_consume('websocket.receive', self._build_message('testmodel', {
            'action': 'retrieve',
            'pk': 1,
            'request_id': 'client-request-id'
        }))
        expected = {
            'action': 'retrieve',
            'data': None,
            'errors': ['Not found.'],
            'response_status': 404,
            'request_id': 'client-request-id'
        }
        self.assertEqual(json_content['payload'], expected)

    def test_retrieve_invalid_pk_404(self):
        json_content = self._send_and_consume('websocket.receive', self._build_message('testmodel', {
            'action': 'retrieve',
            'pk': 'invalid-pk-value',
            'request_id': 'client-request-id'
        }))
        expected = {
            'action': 'retrieve',
            'data': None,
            'errors': ['Not found.'],
            'response_status': 404,
            'request_id': 'client-request-id'
        }
        self.assertEqual(json_content['payload'], expected)

    def test_subscribe(self):

        json_content = self._send_and_consume('websocket.receive', self._build_message("testmodel",{
            'action': 'subscribe',
            'data': {
                'action': 'create'
            },
            'request_id': 'client-request-id'
        }))

        expected_response = {
            'action': 'subscribe',
            'request_id': 'client-request-id',
            'data': {
                'action': 'create'
             },
            'errors': [],
            'response_status': 200
        }

        self.assertEqual(json_content['payload'], expected_response)

        # it should be on the create group
        instance = TestModel.objects.create(name='test-name')

        expected = {
            'action': 'create',
            'data': TestModelSerializer(instance).data,
            'model': 'tests.testmodel',
            'pk': instance.id
        }

        actual = self._get_next_message()

        self.assertEqual(expected, actual['payload'])

    def test_subscribe_failure(self):

        json_content = self._send_and_consume('websocket.receive', self._build_message('testmodel', {
            'action': 'subscribe',
            'data': {
            },
            'request_id': 'client-request-id'
        }))

        expected = {
            'action': 'subscribe',
            'data': None,
            'errors': ['action required'],
            'request_id': 'client-request-id',
            'response_status': 400
        }
        self.assertEqual(expected, json_content['payload'])

    def test_update(self):
        instance = TestModel.objects.create(name='some-test')

        json_content = self._send_and_consume('websocket.receive', self._build_message('testmodel',{
            'action': 'update',
            'pk': instance.id,
            'data': {'name': 'some-value'},
            'request_id': 'client-request-id'
        }))

        instance.refresh_from_db()

        expected = {
            'action': 'update',
            'errors': [],
            'data': TestModelSerializer(instance).data,
            'response_status': 200,
            'request_id': 'client-request-id'
        }

        self.assertEqual(json_content['payload'], expected)

    def test_update_failure(self):
        instance = TestModel.objects.create(name='some-test')

        json_content = self._send_and_consume('websocket.receive', self._build_message('testmodel', {
            'action': 'update',
            'pk': -1,
            'data': {'name': 'some-value'},
            'request_id': 'client-request-id'
        }))

        expected = {
            'data': None,
            'action': 'update',
            'errors': ['Not found.'],
            'response_status': 404,
            'request_id': 'client-request-id'
        }

        self.assertEqual(json_content['payload'], expected)

    def test_no_permission(self):
        instance = TestModel.objects.create(name='some-test')

        """Simulating the connection opening"""
        self._send_and_consume('websocket.connect', {"text": json.dumps({}), "path": "/"})

        json_content = self._send_and_consume('websocket.receive', self._build_message('user:testmodel',{
            'action': 'update',
            'pk': instance.id,
            'data': {'name': 'some-value'},
            'request_id': 'client-request-id'
        }))

        expected = {
            'data': None,
            'action': 'update',
            'errors': ['Permission Denied'],
            'response_status': 401,
            'request_id': None
        }

        self.assertEqual(json_content['payload'], expected)

    def test_permission(self):
        instance = TestModel.objects.create(name='some-test')

        User.objects.create_user('john', 'john@snowmanlabs.com', 'johnpassword')

        """Simulating the connection opening by sending username and password """
        self._send_and_consume('websocket.connect', self._build_authentication('john', 'johnpassword'))

        json_content = self._send_and_consume('websocket.receive', self._build_message('user:testmodel',{
            'action': 'update',
            'pk': instance.id,
            'data': {'name': 'some-value'},
            'request_id': 'client-request-id'
        }))

        """Simulating the disconnection opening """
        self._send_and_consume('websocket.disconnect', self._build_authentication('', ''))

        instance.refresh_from_db()

        expected = {
            'action': 'update',
            'errors': [],
            'data': TestModelSerializer(instance).data,
            'response_status': 200,
            'request_id': 'client-request-id'
        }

        self.assertEqual(json_content['payload'], expected)




channels_api/routing.py
from django.core.serializers.json import json
from channels.generic.websockets import WebsocketDemultiplexer, WebsocketMultiplexer
from channels import Group
from channels.sessions import channel_session
from urllib import parse
from django.contrib.auth import authenticate
from django.conf import settings
from django.contrib.auth.models import User


class Routing(WebsocketDemultiplexer):
    def receive(self, content, **kwargs):
        """Forward messages to all consumers."""
        # Check the frame looks good
        if isinstance(content, dict) and "stream" in content and "payload" in content:
            # Match it to a channel
            for stream, consumer in self.consumers.items():
                if stream == content['stream']:
                    # Extract payload and add in reply_channel
                    payload = content['payload']
                    if not isinstance(payload, dict):
                        raise ValueError("Multiplexed frame payload is not a dict")
                    # The json consumer expects serialized JSON
                    self.message.content['text'] = json.dumps(payload)
                    # Send demultiplexer to the consumer, to be able to answer
                    kwargs['multiplexer'] = WebsocketMultiplexer(stream, self.message.reply_channel)
                    set_channel_session(message=self.message)
                    # Dispatch message
                    consumer(self.message, **kwargs)
                    return

            raise ValueError("Invalid multiplexed frame received (stream not mapped)")
        else:
            raise ValueError("Invalid multiplexed **frame received (no channel/payload key)")

    def connect(self, message, **kwargs):
        set_channel_session(message=self.message)

        username, password = get_username_and_password(message)
        user = authenticate(username=username, password=password)
        if user:
            message.channel_session['_auth_user_id'] = user.pk
            message.channel_session['_auth_user_hash'] = user.get_session_auth_hash()
            message.channel_session['_auth_user_backend'] = 'django.contrib.auth.backends.ModelBackend'

        self.message.reply_channel.send({"accept": True})
        for stream, consumer in self.consumers.items():
            kwargs['multiplexer'] = WebsocketMultiplexer(stream, self.message.reply_channel)
            consumer(message, **kwargs)


@channel_session
def set_channel_session(message):
    pass


def get_username_and_password(message):
    username = message.content.get('username', None)
    password = message.content.get('password', None)
    if username is None:
        if 'query_string' in message:
            query = parse.parse_qs(message['query_string'])
            if 'username' not in query:
                return None, None
            username = query['username'][0]
            password = query['password'][0]
    return username, password


channels_api/bindings.py
import json

from channels.binding import websockets
from channels.binding.base import CREATE, UPDATE, DELETE, BindingMetaclass
from django.http import Http404
from django.utils import six

from rest_framework.exceptions import APIException, NotFound
from rest_framework.generics import get_object_or_404

from .mixins import SerializerMixin, SubscribeModelMixin, CreateModelMixin, UpdateModelMixin, \
    PatchModelMixin, RetrieveModelMixin, ListModelMixin, DeleteModelMixin
from .settings import api_settings


class ResourceBindingMetaclass(BindingMetaclass):
    """
    Metaclass that records action methods
    """

    def __new__(cls, name, bases, body):
        binding = super(ResourceBindingMetaclass, cls).__new__(cls, name, bases, body)

        binding.available_actions = {}
        for methodname in dir(binding):
            attr = getattr(binding, methodname)
            is_action = getattr(attr, 'action', False)
            if is_action:
                kwargs = getattr(attr, 'kwargs', {})
                name = kwargs.get('name', methodname)
                binding.available_actions[name] = methodname

        return binding


@six.add_metaclass(ResourceBindingMetaclass)
class ResourceBindingBase(SerializerMixin, websockets.WebsocketBinding):

    fields = []  # hack to pass cls.register() without ValueError
    queryset = None
    # mark as abstract
    model = None
    serializer_class = None
    lookup_field = 'pk'
    permission_classes = ()

    def deserialize(self, message):
        body = json.loads(message['text'])
        self.request_id = body.get("request_id")
        action = body['action']
        pk = body.get('pk', None)
        data = body.get('data', None)
        return action, pk, data

    @classmethod
    def pre_change_receiver(cls, instance, action):
        """
        Entry point for triggering the binding from save signals.
        """
        if action == CREATE:
            group_names = set()
        else:
            group_names = set(cls.group_names(instance, action))

        if not hasattr(instance, '_binding_group_names'):
            instance._binding_group_names = {}
        instance._binding_group_names[cls] = group_names

    @classmethod
    def post_change_receiver(cls, instance, action, **kwargs):
        """
        Triggers the binding to possibly send to its group.
        """
        old_group_names = instance._binding_group_names[cls]
        if action == DELETE:
            new_group_names = set()
        else:
            new_group_names = set(cls.group_names(instance, action))

        # if post delete, new_group_names should be []
        self = cls()
        self.instance = instance

        # Django DDP had used the ordering of DELETE, UPDATE then CREATE for good reasons.
        self.send_messages(instance, old_group_names - new_group_names, DELETE, **kwargs)
        self.send_messages(instance, old_group_names & new_group_names, UPDATE, **kwargs)
        self.send_messages(instance, new_group_names - old_group_names, CREATE, **kwargs)

    @classmethod
    def group_names(cls, instance, action):
        self = cls()
        groups = [self._group_name(action)]
        if instance.pk:
            groups.append(self._group_name(action, id=instance.pk))
        return groups

    def _group_name(self, action, id=None):
        """Formatting helper for group names."""
        if id:
            return "{}-{}-{}".format(self.model_label, action, id)
        else:
            return "{}-{}".format(self.model_label, action)

    def has_permission(self, user, action, pk):
        check_all = api_settings.CHECK_ALL_PERMISSIONS

        # check that all permissions return True:
        if check_all:
            if self.permission_classes:
                permissions = self.permission_classes
            else:
                permissions = api_settings.DEFAULT_PERMISSION_CLASSES
            for cls in permissions:
                if not cls().has_permission(user, action, pk):
                    return False
            return True
        else:  # check if at least one permission returns True
            for cls in api_settings.DEFAULT_PERMISSION_CLASSES:
                if cls().has_permission(user, action, pk):
                    return True

            if self.permission_classes:
                for cls in self.permission_classes:
                    if cls().has_permission(user, action, pk):
                        return True
            return False

    def filter_queryset(self, queryset):
        return queryset

    def _format_errors(self, errors):
        if isinstance(errors, list):
            return errors
        elif isinstance(errors, six.string_types):
            return [errors]
        elif isinstance(errors, dict):
            return [errors]

    def get_object_or_404(self, pk):
        queryset = self.filter_queryset(self.get_queryset())
        filter_kwargs = {self.lookup_field: pk}
        try:
            return get_object_or_404(queryset, **filter_kwargs)
        except Http404:
            # transform Http404 into an APIException
            raise NotFound

    def get_queryset(self):
        assert self.queryset is not None, (
            "'%s' should either include a `queryset` attribute, "
            "or override the `get_queryset()` method."
            % self.__class__.__name__
        )
        return self.queryset.all()

    def run_action(self, action, pk, data):
        try:
            if not self.has_permission(self.user, action, pk):
                self.reply(action, errors=['Permission Denied'], status=401,
                           request_id=self.request_id)
            elif action not in self.available_actions:
                self.reply(action, errors=['Invalid Action'], status=400,
                           request_id=self.request_id)
            else:
                methodname = self.available_actions[action]
                method = getattr(self, methodname)
                detail = getattr(method, 'detail', True)
                if detail:
                    rv = method(pk, data=data)
                else:
                    rv = method(data=data)
                data, status = rv
                self.reply(action, data=data, status=status, request_id=self.request_id)
        except APIException as ex:
            self.reply(action, errors=self._format_errors(ex.detail), status=ex.status_code, request_id=self.request_id)

    def reply(self, action, data=None, errors=[], status=200, request_id=None):
        """
        Helper method to send a encoded response to the message's reply_channel.
        """
        payload = {
            'errors': errors,
            'data': data,
            'action': action,
            'response_status': status,
            'request_id': request_id
        }
        return self.message.reply_channel.send(self.encode(self.stream, payload))


class ResourceBinding(CreateModelMixin, RetrieveModelMixin, ListModelMixin,
    UpdateModelMixin, PatchModelMixin, DeleteModelMixin, SubscribeModelMixin, ResourceBindingBase):

    # mark as abstract
    model = None


class ReadOnlyResourceBinding(RetrieveModelMixin, ListModelMixin,
    ResourceBindingBase):

    # mark as abstract
    model = None
